### PR TITLE
load network from localstorage + configuration page

### PR DESCRIPTION
**Pull request by mk-emeks**
# New stuff

- recoil for state management
- configuration page
- network form
- redirect if localstorage is empty

# UI
Screenshot(9)

## Ref
https://github.com/emeks-studio/sensum/issues/36

### CODE REVIEW

File: sensum-pwa/src/app/App.res [hunk 1]
```diff
@@ -44,6 +47,7 @@ let make = () => {
     Promise.resolve()
   })
   ->ignore
-
+  
+  State.Configuration.loadNetworkHook()
```
Comment (mk-emeks) [inline sensum-pwa/src/app/App.res:51]: start app by loading settings from local storage

File: sensum-pwa/src/router/Router.res [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
 let make = () => {
+  let networkUrl = Recoil.useRecoilValue(State.Configuration.networkUrlAtom)
   let url = RescriptReactRouter.useUrl()
-  
-  switch url.path {
-    | list{} => <Pages_Sensations/>
-    | list{"sensations"} => <Pages_Sensations/>
-    | _ => <Pages_NotFound/>
+
+  switch networkUrl {
```
Comment (mk-emeks) [inline sensum-pwa/src/router/Router.res:7]: if no network url, render configuration page!

File: sensum-pwa/src/app/App.res (hunk 1 above)
Comment (randomeduc) [reply sensum-pwa/src/app/App.res:51]: This file feels weird jaja, guess it's because of the syntax, dunno

Comment (mirko-plowtech) [reply sensum-pwa/src/app/App.res:51]: Don't worry, the file still has a lot of things that are wrong and will be removed. Like the `Sensations` and blockchain related stuff

File: sensum-pwa/src/Main.res [hunk 3]
```diff
@@ -3,6 +3,6 @@
 let rootQuery = ReactDOM.querySelector("#root")
 
 switch rootQuery {
-| Some(root) => ReactDOM.render(<App />, root)
+| Some(root) => ReactDOM.render(<Recoil.RecoilRoot> <App /> </Recoil.RecoilRoot>, root)
```
[the remark this answers is not in the document]
Comment (mirko-plowtech) [reply sensum-pwa/src/Main.res:6]: Yes and no. I mean, is the Recoil provider, requested by `recoil` library: https://recoiljs.org/docs/api-reference/core/RecoilRoot/#example

^ As you can see in the example, is similar in plain ReactJS

File: sensum-pwa/src/state/configuration/State_Configuration.res [hunk 4]
```diff
@@ -18,7 +18,7 @@ let loadNetworkHook = () => {
   })
 }
 
-let useNetworkHook = () => {
+let useNetwork = () => {
```
Comment (mk-emeks) [inline sensum-pwa/src/state/configuration/State_Configuration.res:21]: prefix `use` for any custom hook

Comment (mk-emeks) [reply sensum-pwa/src/state/configuration/State_Configuration.res:21]: (just a convention)

File: sensum-pwa/src/state/configuration/State_Configuration.res [hunk 5]
```diff
@@ -9,7 +9,7 @@ let networkUrlAtom = Recoil.atom({
   default: initialUrl,
 })
 
-let loadNetworkHook = () => {
+let useEffectLoadNetworkFromStorage = () => {
```
Comment (mk-emeks) [inline sensum-pwa/src/state/configuration/State_Configuration.res:12]: prefix `useEffect` for any custom hook effect

File: sensum-pwa/src/blockchain/eth/Ethers.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { ethers } from 'ethers';
+
+export async function getNetwork(networkUrl) {
+  const provider = new ethers.providers.JsonRpcProvider(networkUrl);
```
Comment (mk-emeks) [inline sensum-pwa/src/blockchain/eth/Ethers.js:4]: IMHO, is better to repeat this line than try to create a binding for the provider (given `ethers` is a stateful library)

File: sensum-pwa/src/blockchain/eth/Sensations.res [hunk 7]
```diff
@@ -27,3 +27,38 @@ external getSensationsLength: () => Js.Promise.t<bigInt> = "getSensationsLength"
 
 @module("./Sensations.js")
 external newSensation: (~s: sensation) => Js.Promise.t<confirmedTransaction> = "newSensation"
+
+// Ex. get sensation
```
Comment (mk-emeks) [inline sensum-pwa/src/blockchain/eth/Sensations.res:31]: Eventually I will remove this!

File: sensum-pwa/src/router/Router.res [hunk 8]
```diff
@@ -1,13 +1,44 @@
 // Ref. https://rescript-lang.org/docs/react/latest/router#basic-example
 @react.component
 let make = () => {
-  let networkUrl = Recoil.useRecoilValue(State.Configuration.networkUrlAtom)
-  let url = RescriptReactRouter.useUrl()
+  // Notice that for reads is ok/safe to just use directly the atom.
+  // But for writes, we should use some of the exposed hooks.
+  let mNetworkUrl = Recoil.useRecoilValue(State.Configuration.networkUrlAtom)
```
Comment (mk-emeks) [inline sensum-pwa/src/router/Router.res:6]: convetion; `mVariableName` means, maybe variable name

**Pull request merged**